hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/utils/HoodieHiveUtils.java

```diff
@@ -146,4 +155,99 @@ public static List<String> getIncrementalTableNames(JobContext job) {
   public static boolean isIncrementalUseDatabase(Configuration conf) {
     return conf.getBoolean(HOODIE_INCREMENTAL_USE_DATABASE, false);
   }
+
+  public static final String HIVE_TIMESTAMP_TYPE_CLASS = "org.apache.hadoop.hive.common.type.Timestamp";
+  public static final String TIMESTAMP_WRITEABLE_V2_CLASS = "org.apache.hadoop.hive.serde2.io.TimestampWritableV2";
+  public static final boolean SUPPORT_TIMESTAMP_WRITEABLE_V2;
+  private static final Class TIMESTAMP_CLASS;
+  private static final Method SET_TIME_IN_MILLIS;
+  private static final Constructor TIMESTAMP_WRITEABLE_V2_CONSTRUCTOR;
+
+  public static final String DATE_WRITEABLE_V2_CLASS = "org.apache.hadoop.hive.serde2.io.DateWritableV2";
+  public static final boolean SUPPORT_DATE_WRITEABLE_V2;
+  private static final Constructor DATE_WRITEABLE_V2_CONSTRUCTOR;
+
+  static {
+    // timestamp
+    Option<ImmutableTriple<Class, Method, Constructor>> timestampTriple = Option.ofNullable(() -> {
+      try {
+        Class timestampClass = Class.forName(HIVE_TIMESTAMP_TYPE_CLASS);
+        Method setTimeInMillis = timestampClass.getDeclaredMethod("setTimeInMillis", long.class);
+        Class twV2Class = Class.forName(TIMESTAMP_WRITEABLE_V2_CLASS);
+        return ImmutableTriple.of(timestampClass, setTimeInMillis, twV2Class.getConstructor(timestampClass));
+      } catch (ClassNotFoundException | NoSuchMethodException e) {
+        LOG.trace("can not find hive3 timestampv2 class or method, use hive2 class!", e);
+        return null;
+      }
+    });
+    SUPPORT_TIMESTAMP_WRITEABLE_V2 = timestampTriple.isPresent();
+    if (SUPPORT_TIMESTAMP_WRITEABLE_V2) {
+      LOG.trace("use org.apache.hadoop.hive.serde2.io.TimestampWritableV2 to read hudi timestamp columns");
+      ImmutableTriple<Class, Method, Constructor> triple = timestampTriple.get();
+      TIMESTAMP_CLASS = triple.left;
+      SET_TIME_IN_MILLIS = triple.middle;
+      TIMESTAMP_WRITEABLE_V2_CONSTRUCTOR = triple.right;
+    } else {
+      LOG.trace("use org.apache.hadoop.hive.serde2.io.TimestampWritable to read hudi timestamp columns");
+      TIMESTAMP_CLASS = null;
+      SET_TIME_IN_MILLIS = null;
+      TIMESTAMP_WRITEABLE_V2_CONSTRUCTOR = null;
+    }
+
+    // date
+    Option<Constructor> dateConstructor = Option.ofNullable(() -> {
```

Comment:
please re-work on this API; you should make use of `orElseGet()` to work with supplier